### PR TITLE
Bug fix #991

### LINKED ISSUE #991: Missing AWS - RESOURCE -> AWSPrincipal
*Description*:
> What issue is being seen?  When ingesting policy, we create AWSPrincipal but we don't create and associate it to it's respective AWSAccount. The AWSPrincipal can be in the account being sync but can also be part of another account that is trusted.

https://github.com/lyft/cartography/blob/8db4a7eb1b57df8e681eeef09523a34e963aa09b/cartography/intel/aws/iam.py#L297
    ingest_policy_statement = """
    MERGE (spnnode:AWSPrincipal{arn: {SpnArn}})
    ON CREATE SET spnnode.firstseen = timestamp()
    SET spnnode.lastupdated = {aws_update_tag}, spnnode.type = {SpnType}
    WITH spnnode
    MATCH (role:AWSRole{arn: {RoleArn}})
    MERGE (role)-[r:TRUSTS_AWS_PRINCIPAL]->(spnnode)
    ON CREATE SET r.firstseen = timestamp()
    SET r.lastupdated = {aws_update_tag}
    """

*To Reproduce*:
Create a trust relationship to another AWS account that is not part of the scanning. Even if the AWSPrincipal is part of the scanned account, the RESOURCE won't be created (see code)

*Logs*:
N/A - no error is generated

*Screenshots*:
> If applicable, add screenshots to help explain your problem.

*Please complete the following information:*:
>  - Cartography release version : 0.65.0
>  - Python version: [e.g. 3.7.4] 3.10

### PR DESCRIPTION
**Pull request by sachafaust**

### CODE REVIEW

File: cartography/data/jobs/analysis/aws_foreign_accounts.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
+  "statements": [
+    {
+      "__comment": "This analyze AWS accounts we created and tag the ones that are forein. Forein accounts are ones that were not in the sync scope",
```
Comment (trodery) [inline cartography/data/jobs/analysis/aws_foreign_accounts.json:4]: I believe the spelling of `forein` should be `foreign`

File: cartography/intel/aws/iam.py [hunk 2]
```diff
@@ -739,3 +767,21 @@ def sync(
...
+@timeit
+def get_account_from_arn(arn: str) -> str:
+    # ARN documentation
+    # https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html
+
+    if not arn:
+        return ""
+
+    if not arn.startswith("arn:"):
+        return ""
```
[the remark this answers is not in the document]
Comment (sachafaust) [reply cartography/intel/aws/iam.py:781]: core AWS service principals do no have ARN - example - ec2.amazonaws.com

We test for this in various places - [](https://github.com/lyft/cartography/blob/1960113953c436bf8dca4098c5336fde19c3925f/tests/integration/cartography/intel/aws/test_iam.py#L99)

[the remark this answers is not in the document]
Comment (ramonpetgrave64) [reply cartography/intel/aws/iam.py:781]: Thanks. But I'm worried about the case `if not arn:`. It seems like that should never be `True`.

[the remark this answers is not in the document]
Comment (sachafaust) [reply cartography/intel/aws/iam.py]: I can't come up with example for that and removed it.

File: cartography/intel/aws/ec2/vpc_peerings.py [hunk 3]
```diff
@@ -51,12 +51,12 @@ def load_vpc_peerings(
     SET rvpc.lastupdated = $update_tag, rvpc.vpcid = vpc_peering.RequesterVpcInfo.VpcId
 
     MERGE (aaccount:AWSAccount{id: vpc_peering.AccepterVpcInfo.OwnerId})
-    ON CREATE SET aaccount.firstseen = timestamp(), aaccount.foreign = true
-    SET aaccount.lastupdated = $update_tag
+    ON CREATE SET aaccount.firstseen = timestamp()
+    SET aaccount.lastupdated = {update_tag}
```
Comment (ramonpetgrave64) [inline cartography/intel/aws/ec2/vpc_peerings.py:55]:
```suggestion
    SET aaccount.lastupdated = $update_tag
```

File: cartography/intel/aws/ec2/vpc_peerings.py [hunk 4]
```diff
@@ -51,12 +51,12 @@ def load_vpc_peerings(
...
 
     MERGE (raccount:AWSAccount{id: vpc_peering.RequesterVpcInfo.OwnerId})
-    ON CREATE SET raccount.firstseen = timestamp(), raccount.foreign = true
-    SET raccount.lastupdated = $update_tag
+    ON CREATE SET raccount.firstseen = timestamp()
+    SET raccount.lastupdated = {update_tag}
```
Comment (ramonpetgrave64) [inline cartography/intel/aws/ec2/vpc_peerings.py:59]:
```suggestion
    SET raccount.lastupdated = $update_tag
```

File: cartography/intel/aws/iam.py [hunk 5]
```diff
@@ -739,3 +767,18 @@ def sync(
...
     )
+
+
+@timeit
+def get_account_from_arn(arn: str) -> str:
+    # ARN documentation
+    # https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html
+
+    if not arn.startswith("arn:"):
+        return ""
```
Comment (ramonpetgrave64) [inline cartography/intel/aws/iam.py:778]:
```suggestion
        # must be a service principal arn, such as ec2.amazonaws.com
        return ""
```

File: cartography/data/jobs/cleanup/aws_import_principals_cleanup.json [hunk 6]
```diff
@@ -21,11 +21,10 @@
       "iterationsize": 100
     },
     {
-    "query": "MATCH (:AWSAccount{id: $AWS_ID})-[:RESOURCE]->(n:AWSPrincipal) WHERE n.lastupdated <> $UPDATE_TAG WITH n LIMIT $LIMIT_SIZE DETACH DELETE (n)",
+    "query": "MATCH (n:AWSAccount{foreign:true})-[:RESOURCE]->(n:AWSPrincipal) WHERE n.lastupdated <> {UPDATE_TAG} WITH n LIMIT {LIMIT_SIZE} DETACH DELETE (n)",
```
Comment (ramonpetgrave64) [inline cartography/data/jobs/cleanup/aws_import_principals_cleanup.json:24]: I know the code and cleanup jobs are messy, and so I'm afraid to delete this original query. But we should add your new query.

File: cartography/intel/aws/organizations.py [hunk 7]
```diff
@@ -92,7 +92,7 @@ def load_aws_accounts(
 ) -> None:
     query = """
     MERGE (aa:AWSAccount{id: $ACCOUNT_ID})
-    ON CREATE SET aa.firstseen = timestamp()
+    ON CREATE SET aa.firstseen = timestamp(), aa.inscope=true
```
Comment (ramonpetgrave64) [inline cartography/intel/aws/organizations.py:95]: We should maybe also be intentional to delete the `foreign` property if it exists.

File: cartography/intel/aws/ec2/network_interfaces.py [hunk 8]
```diff
@@ -82,9 +82,8 @@ def load_network_interfaces(
             ON CREATE SET r.firstseen = timestamp()
             SET r.lastupdated = $update_tag
         WITH network_interface, netinf
-
         MERGE (acc:AWSAccount{id: $aws_account_id})
-        ON CREATE SET acc.firstseen = timestamp()
+        ON CREATE SET acc.firstseen = timestamp(), acc.inscope=true
```
Comment (ramonpetgrave64) [inline cartography/intel/aws/ec2/network_interfaces.py:86]: We should maybe also be intentional to delete the `foreign` property if it exists.

File: cartography/data/jobs/analysis/aws_foreign_accounts.json [hunk 9]
```diff
@@ -0,0 +1,15 @@
+{
+  "statements": [
+    {
+      "__comment": "This analyze AWS accounts we created and tag the ones that are foreign. Foreign accounts are ones that were not in the sync scope",
+      "query": "MATCH (foreign:AWSAccount) where NOT EXISTS(foreign.inscope) SET foreign.foreign = true",
+      "iterative": false
+    },
+    {
+      "__comment": "Remove accounts that were set with foreign and inscope. This can happen as we finish the list of sync accounts through assume role mapping and vpc peering",
```
Comment (ramonpetgrave64) [inline cartography/data/jobs/analysis/aws_foreign_accounts.json:9]: Unlikely scenario, but what about when an account is now `foreign`, but used to be `inscope`? This query would incorrectly set the account as `inscope`.

We should maybe also be intentional to delete the `foreign` property if it exists.

File: cartography/data/jobs/cleanup/aws_import_principals_cleanup.json (hunk 6 above)
Comment (sachafaust) [reply cartography/data/jobs/cleanup/aws_import_principals_cleanup.json:24]: Added comment on why we don't scope the query to $AWS_ID.

Comment (sachafaust) [reply cartography/data/jobs/cleanup/aws_import_principals_cleanup.json:24]: Fixed dupe n variable.

File: cartography/intel/aws/ec2/network_interfaces.py (hunk 8 above)
Comment (sachafaust) [reply cartography/intel/aws/ec2/network_interfaces.py:86]: Fixed in aws_account_cleanup.json job

File: cartography/data/jobs/analysis/aws_foreign_accounts.json (hunk 9 above)
Comment (sachafaust) [reply cartography/data/jobs/analysis/aws_foreign_accounts.json:9]: Case 1) Removal of foreign is done in aws_account_cleanup.json

Case 2) For the cases "an account is now foreign, but used to be inscope? " There might be other cases but I can see this happening if the user change profile and runs a scan on top of existing graph + the old AWS accounts are trusted by the new set of AWS account in the new profile.

We can wipe the inscope + foreign across all accounts at the beginning of the AWS module - That would impact anything that runs analysis using both foreign and inscope attributes during the full sync.

@ramonpetgrave64 Any other suggestions? Is it a blocker?

File: cartography/intel/aws/organizations.py (hunk 7 above)
Comment (sachafaust) [reply cartography/intel/aws/organizations.py:95]: fixed in aws_foreign.accounts.json

File: cartography/intel/aws/organizations.py [hunk 10]
```diff
@@ -92,7 +92,7 @@ def load_aws_accounts(
...
     SET aa.lastupdated = $aws_update_tag, aa.name = $ACCOUNT_NAME
```
Comment (ramonpetgrave64) [inline cartography/intel/aws/organizations.py:96]:
```suggestion
    ON CREATE SET aa.firstseen = timestamp()
    SET aa.lastupdated = $aws_update_tag, aa.name = $ACCOUNT_NAME, aa.inscope=true
    REMOVE aa.foreign
```

File: cartography/data/jobs/analysis/aws_foreign_accounts.json (hunk 9 above)
Comment (ramonpetgrave64) [reply cartography/data/jobs/analysis/aws_foreign_accounts.json:9]: I think my suggestion below in organizations.py covers the case. Let me know if that works.

**Review by ramonpetgrave64 — CHANGES_REQUESTED** [review]

**Review by ramonpetgrave64 — APPROVED** [review]
I've tested the changes and they work well. But I'll gold off merging for another day, if you have another idea about dealing with the edge cases.

**Comment by sachafaust** [comment]
I'm good with current state, the edge case remaining on my end is one where the user change profile and AWS account that use to be inscope is not foreign only.

**Pull request merged**